client/src/components/graph/graph.tsx

```diff
@@ -557,6 +538,122 @@ class Graph extends React.Component<GraphProps, GraphState> {
     if (selectionTool === "lasso") this.handleLassoDeselectAction();
   }
 
+  async handleImageDownload(regl: GraphState["regl"]) {
+    const { dispatch, screenCap, mountCapture, layoutChoice, colors } =
+      this.props;
+
+    if (
+      !this.reglCanvas ||
+      !screenCap ||
+      !regl ||
+      this.isDownloadingImage ||
+      !this.openseadragon
+    ) {
+      return;
+    }
+
+    const { width, height } = this.reglCanvas;
+
+    /**
+     * (thuang): This prevents re-rendering causes a second image download
+     */
+    this.isDownloadingImage = true;
+
+    const imageCanvas = this.openseadragon.drawer.canvas as HTMLCanvasElement;
+    const graphCanvas = regl._gl.canvas as HTMLCanvasElement;
+
+    // Create an offscreen canvas
+    const offscreenCanvas = document.createElement("canvas");
+    offscreenCanvas.width = width;
+    offscreenCanvas.height = height;
+
+    const ctx = offscreenCanvas.getContext("2d");
+
+    if (!ctx) {
+      console.error("Failed to get 2D context for the offscreen canvas");
+      this.isDownloadingImage = false;
+      return;
+    }
+
+    try {
+      // Convert both canvases to data URLs
+      const imageDataURL = imageCanvas.toDataURL();
+      const graphDataURL = graphCanvas.toDataURL();
+
+      // Load both data URLs into image objects
+      const [image, graphImage] = await Promise.all([
+        loadImage(imageDataURL),
+        loadImage(graphDataURL),
+      ]);
+
+      // Calculate aspect ratio
+      const aspectRatio = image.width / image.height;
+      let targetWidth;
+      let targetHeight;
+      let offsetX = 0;
+      let offsetY = 0;
+
+      if (offscreenCanvas.width / offscreenCanvas.height > aspectRatio) {
+        // Fit by height
+        targetHeight = offscreenCanvas.height;
+        targetWidth = targetHeight * aspectRatio;
+        offsetX = (offscreenCanvas.width - targetWidth) / 2;
+      } else {
+        // Fit by width
+        targetWidth = offscreenCanvas.width;
+        targetHeight = targetWidth / aspectRatio;
+        offsetY = (offscreenCanvas.height - targetHeight) / 2;
+      }
+
+      // Draw the OpenSeadragon image as background with proper scaling
+      ctx.drawImage(image, offsetX, offsetY, targetWidth, targetHeight);
+
+      // Draw the graph image on top, ensuring it covers the entire canvas
+      ctx.drawImage(
+        graphImage,
+        0,
+        0,
+        offscreenCanvas.width,
+        offscreenCanvas.height
+      );
+
+      const graphImageURI = offscreenCanvas.toDataURL("image/png");
+
+      if (mountCapture) {
+        this.setState({ testImageSrc: graphImageURI });
+        dispatch({ type: "test: screencap end" });
+      } else {
+        downloadImage(graphImageURI, layoutChoice);
+
+        let categoricalLegendImageURI: string | null = null;
+
+        // without this, the legend is drawn offscreen
+        const PADDING_PX = 100;
```

Comment:
Does the canvas width and height change according to the user's screen resolution and window resize state? If so does it make sense for this to be a percentage value instead of arbitrary value?

Reply:
ooh this is a great question! Since @seve added it originally let's wait for him to answer on this one. But my gut feeling is that since this constant only affects the legend, which occupies the same width regardless of viewport size, so I'm thinking it should be okay

I'll merge the PR first and then submit a subsequent PR to fix if Seve thinks we need to update anything

Thanks so much for the quick review!